htdocs/core/class/html.form.class.php

```diff
@@ -1572,126 +1572,262 @@ public function select_remises($selected, $htmlname, $filter, $socid, $maxvalue
+	 * @param	string	$selected   			Id contact pre-selectionne
+	 * @param	string	$htmlname  	    		Name of HTML field ('none' for a not editable field)
+	 * @param	int		$showempty      		0=no empty value, 1=add an empty value, 2=add line 'Internal' (used by user edit), 3=add an empty value only if more than one record into list
+	 * @param	string	$exclude        		List of contacts id to exclude
+	 * @param	string	$limitto				Disable answers that are not id in this array list
+	 * @param	integer	$showfunction   		Add function into label
+	 * @param	string	$moreclass				Add more class to class style
+	 * @param	integer	$showsoc	    		Add company into label
+	 * @param	int		$forcecombo				Force to use combo box
+	 * @param	array	$events					Event options. Example: array(array('method'=>'getContacts', 'url'=>dol_buildpath('/core/ajax/contacts.php',1), 'htmlname'=>'contactid', 'params'=>array('add-customer-contact'=>'disabled')))
+	 * @param	bool	$options_only			Return options only (for ajax treatment)
+	 * @param	string	$moreparam				Add more parameters onto the select tag. For example 'style="width: 95%"' to avoid select2 component to go over parent container
+	 * @param	string	$htmlid					Html id to use instead of htmlname
+	 * @param	bool	$multiple				add [] in the name of element and add 'multiple' attribut
+	 * @param	int		$disableifempty			Set tag 'disabled' on select if there is no choice
+	 * @param	bool	$new_ajax				Enabled new ajax search method
+	 * @param	string	$thirdparty_htmlname	Name of HTML field of the linked thirparty input
+	 * @param	int		$limit					Maximum number of elements
+	 * @param	int		$hidelabel				Hide label (0=no, 1=yes, 2=show search icon (before) and placeholder, 3 search icon after)
+	 * @param	string	$ajaxurl				Url for ajax request in ajax_combobox()
+	 * @param	array	$ajaxparameters			Additionnals parameters for ajax request in ajax_combobox()
+	 * @param	array	$ajaxoptions			Options for ajax_autocompleter // Not used for this moment todo add ajaxoptions to ajax_combobox()
+	 * @return	int								<0 if KO, Nb of contact in list if OK
+	 * @deprected								You can use selectcontacts directly (warning order of param was changed)
 	 */
-	public function select_contacts($socid, $selected = '', $htmlname = 'contactid', $showempty = 0, $exclude = '', $limitto = '', $showfunction = 0, $moreclass = '', $showsoc = 0, $forcecombo = 0, $events = array(), $options_only = false, $moreparam = '', $htmlid = '')
+	public function select_contacts($socid, $selected = '', $htmlname = 'contactid', $showempty = 0, $exclude = '', $limitto = '', $showfunction = 0, $moreclass = '', $showsoc = 0, $forcecombo = 0, $events = array(), $options_only = false, $moreparam = '', $htmlid = '', $multiple = false, $disableifempty = 0, $new_ajax = false, $thirdparty_htmlname = '', $limit = 20, $hidelabel = 1, $ajaxurl = DOL_URL_ROOT . '/contact/ajax/contact.php', $ajaxparameters = array(), $ajaxoptions = array())
```

Comment:
Is having a dynamic default value based on a runtime var as a default paramater in declaration working on all cases/PHP ? (here $ajaxurl = DOL_URL_ROOT...)
May be you can set default value to '' and into function set it to DOL_URL_ROOT ... if '' ?
if may work, but it looks strange to me.

Reply:
From the manual, since PHP 5.6:

"Example 6 Using non-scalar types as default values (...)
The default value must be a constant expression, not (for example) a variable, a class member or a function call."

So, it seems correct to me to use constant as default value in function.

https://www.php.net/manual/en/functions.arguments.php#functions.arguments.default

Reply:
But that was not always true in PHP 5.4

Reply:
However, this make still code declaration not easy to understand. this may also obfuscate bugs on some Continuous integration like scrutinizer.
Can you replace with $ajaxurl = '' and in code of function make
```
if (empty($ajaxurl)) {
   $ajaxurl = DOL_URL_ROOT . '/contact/ajax/contact.php';
}
```